File: Animal/Animal.cpp
#include "Animal.h"
#include "../Human/Human.h"

Animal::Animal(int strength, int initiative, char symbol, Color color, World *world, int x, int y, int parent1Id, int parent2Id) :
    Organism(strength,
             initiative,
             symbol,
             color,
             world,
             Position(x, y),
             parent1Id,
             parent2Id) { }

void Animal::action() {
  this->setPreviousX(this->getPosition().getX());
  this->setPreviousY(this->getPosition().getY());

  Position newPosition = this->getWorld()->generateNewPosition(this->getPosition().getX(),
                                                               this->getPosition().getY());
  this->setPosition(newPosition.getX(), newPosition.getY());

  this->getWorld()->addLog("%s POS:(%d, %d) moved to (%d, %d)",
                           this->getType().c_str(),
                           this->getPreviousX(),
                           this->getPreviousY(),
                           this->getPosition().getX(),
                           this->getPosition().getY());
}

void Animal::collision(Organism *organism) {
  if (this->getType() == organism->getType()) {
    if (this->isParent(organism) ||
        this->isParent(this) ||
        this->hasSameParents(organism))
      return;

    int x = this->getPosition().getX();
    int y = this->getPosition().getY();

    if (this->getWorld()->getOrganismAtPosition(this, this->getPreviousX(), this->getPreviousY()) == nullptr &&
      this->getWorld()->getOrganismAtPosition(organism, organism->getPreviousX(), organism->getPreviousY()) == nullptr) {

      this->setPosition(this->getPreviousX(), this->getPreviousY());
      organism->setPosition(organism->getPreviousX(), organism->getPreviousY());

      this->getWorld()->addOrganism(this->createChild(x,
                                                      y,
                                                      this->getId(),
                                                      organism->getId()));

      this->getWorld()->addLog("%s POS:(%d, %d) breeds (%d w %d)",
                               this->getType().c_str(),
                               this->getPosition().getX(),
                               this->getPosition().getY(),
                               this->getId(),
                               organism->getId());
    }
  } else if (organism->getInitiative() == 0) {
    this->getWorld()->addLog("%s POS:(%d, %d) ate %s POS:(%d, %d)",
                             this->getType().c_str(),
                             this->getPosition().getX(),
                             this->getPosition().getY(),
                             organism->getType().c_str(),
                             organism->getPosition().getX(),
                             organism->getPosition().getY());

    organism->collision(this);
    organism->setAlive(false);
  } else {
    this->getWorld()->addLog("%s POS:(%d, %d) attacked %s POS:(%d, %d)",
                             this->getType().c_str(),
                             this->getPosition().getX(),
                             this->getPosition().getY(),
                             organism->getType().c_str(),
                             organism->getPosition().getX(),
                             organism->getPosition().getY());

    if (organism->deflectedAttack(this)) {
      if (organism->getPosition().getX() ==  this->getPosition().getX() &&
          organism->getPosition().getY() == this->getPosition().getY() &&
          organism->getType() != "Human") {
        this->getWorld()->addLog("%s POS:(%d, %d) S:%d killed %s POS:(%d, %d) S:%d",
                                 organism->getType().c_str(),
                                 this->getPosition().getX(),
                                 this->getPosition().getY(),
                                 organism->getStrength(),
                                 this->getType().c_str(),
                                 organism->getPosition().getY(),
                                 organism->getPosition().getX(),
                                 this->getStrength());

        this->setAlive(false);
      } else {
        this->getWorld()->addLog("%s POS:(%d, %d) deflected attack from %s POS:(%d, %d)",
                                 organism->getType().c_str(),
                                 organism->getPosition().getX(),
                                 organism->getPosition().getY(),
                                 this->getType().c_str(),
                                 this->getPosition().getX(),
                                 this->getPosition().getY());
      }

    } else {
      this->getWorld()->addLog("%s POS:(%d, %d) S:%d killed %s POS:(%d, %d) S:%d",
                               this->getType().c_str(),
                               this->getPosition().getX(),
                               this->getPosition().getY(),
                               this->getStrength(),
                               organism->getType().c_str(),
                               organism->getPosition().getX(),
                               organism->getPosition().getY(),
                               organism->getStrength());

      organism->setAlive(false);
    }
  }
}

bool Animal::isParent(Organism* organism) const {
  if (organism->getParent1Id() == this->getId() ||
      organism->getParent2Id() == this->getId()) {
    return true;
  }
  return false;
}

bool Animal::hasSameParents(Organism* organism) const {
  if (organism->getParent1Id() == -1 || this->getParent1Id() == -1) return false;

  if ((organism->getParent1Id() == this->getParent1Id() && organism->getParent2Id() == this->getParent2Id()) ||
      (organism->getParent1Id() == this->getParent2Id() && organism->getParent2Id() == this->getParent1Id())) {
    return true;
  }
  return false;
}

File: src/Animal/Animal.java
package Animal;

import Field.Field;
import Organism.*;
import World.World;

public abstract class Animal extends Organism {

    protected Field lastField;

    public Animal(OrganismData data, World world, Field field) {
        super(data, world, field);
        this.lastField = field;
    }

    @Override
    public void action() {
        Field field = this.world.getNeighbourField(this.currentField);

        this.lastField = this.currentField;
        this.currentField.setOrganism(null);
        this.currentField = field;
        this.currentField.setOrganism(this);
    }

    @Override
    public void collision(Organism organism) {
        if (this.getType().equals(organism.getType())) {
            this.reproduce(organism);
        } else if (organism.getData().getInitiative() == 0) {
            this.eatPlant(organism);
        } else {
            this.fight(organism);
        }
    }

    public void moveOrganismToPreviousField() {
        Organism otherOrganism = this.world.getOrganismAtField(this.currentField, this);
        if (otherOrganism != null) {
            this.currentField.setOrganism(otherOrganism);
        }

        this.currentField = lastField;
        this.currentField.setOrganism(this);
    }

    private void reproduce(Organism organism) {
        Animal animal = (Animal) organism;
        Field childField = this.getCurrentField();

        this.moveOrganismToPreviousField();
        animal.moveOrganismToPreviousField();

        if (this.hasSameParents(organism) || this.isParent(organism) || animal.isParent(this)) return;

        this.reproduce(childField, this.getData().getId(), organism.getData().getId());
        this.getWorld().addLog(this.getType() + " (" + this.getCurrentField().getNumber() + ", " + organism.getCurrentField().getNumber() + ") reproduced (" + childField.getNumber() + ")");
     }

    private void eatPlant(Organism organism) {
        this.getWorld().addLog(this.getType() + " (" + this.getCurrentField().getNumber() + ") ate " + organism.getType());
        organism.collision(this);
        organism.getData().setAlive(false);
    }

    private void fight(Organism organism) {
        this.world.addLog(this.getType() + " (" + this.getCurrentField().getNumber() + ") attacked " + organism.getType() + " (" + organism.getCurrentField().getNumber() + ")");

        Animal animal = (Animal) organism;
        if (this.getData().getStrength() >= organism.getData().getStrength()) {
            if (animal.deflectedAttack(this)) {
                this.world.addLog(animal.getType() + " (" + animal.getCurrentField().getNumber() + ") deflected attack");
            } else {
                this.world.addLog(this.getType() + " killed " + animal.getType() + " (" + animal.getCurrentField().getNumber() + ")");
                animal.getData().setAlive(false);
            }
        } else {
            this.world.addLog(animal.getType() + " killed " + this.getType() + " (" + this.getCurrentField().getNumber() + ")");
            this.getData().setAlive(false);
        }
    }

    private boolean hasSameParents(Organism organism) {
        if (this.getData().getParent1Id() == -1 || organism.getData().getParent1Id() == -1) return false;
        return this.getData().getParent1Id() == organism.getData().getParent1Id() || this.getData().getParent2Id() == organism.getData().getParent2Id() || this.getData().getParent1Id() == organism.getData().getParent2Id() || this.getData().getParent2Id() == organism.getData().getParent1Id();
    }

    public boolean isParent(Organism organism) {
        return this.getData().getId() == organism.getData().getParent1Id() || this.getData().getId() == organism.getData().getParent2Id();
    }

    public boolean deflectedAttack(Organism attacker) {
        return attacker.getData().getStrength() < this.getData().getStrength();
    }

    public abstract void reproduce(Field field, int parent1Id, int parent2Id);
}
